packages/flutter_test/lib/src/finders.dart

```diff
@@ -460,57 +469,544 @@ class CommonFinders {
+
+  /// Returns a variant of this finder that only matches the last item
+  /// found by this finder.
+  FinderBase<TCandidate> get last => _LastFinder<TCandidate>(this);
+
+  /// Returns a variant of this finder that only matches the item at the
+  /// given index found by this finder.
+  FinderBase<TCandidate> at(int index) => _IndexFinder<TCandidate>(this, index);
+
+  /// Returns all the items in the given list that match this
+  /// finder's requirements.
+  ///
+  /// When implementing your own finders that inherit directly from
+  /// [FinderBase], this is the main method to override. If your finder
   /// can efficiently be described just in terms of a predicate
-  /// function, consider extending [MatchFinder] instead.
-  Iterable<Element> apply(Iterable<Element> candidates);
+  /// function, consider mixing in [MatchFinderMixin] instead.
+  @protected
+  Iterable<TCandidate> findInCandidates(Iterable<TCandidate> candidates);
 
-  /// Whether this finder skips nodes that are offstage.
+  /// Searches all of the candidates and returns the result of that search.
   ///
-  /// If this is true, then the elements are walked using
+  /// See also:
+  ///
+  ///   * [found] which will return the latest results without re-executing the
+  ///     search.
+  ///   * [tryFind] which will indicate whether any results were found rather
+  ///     than directly returning results.
+  FinderResult<TCandidate> find() {
+    _found = FinderResult<TCandidate>(this.requirements, findInCandidates(allCandidates));
+    return found;
+  }
+
+  /// Searches all of the candidates and returns whether the search found any
+  /// matching candidates.
+  ///
+  /// This is useful in cases where an action needs to be repeated while or
+  /// until a finder has results. The results from the search can be accessed
+  /// using the [found] property without re-executing the search.
+  bool tryFind() {
+    find();
+    return found.isNotEmpty;
+  }
+
+  /// A string representation of this finder or its results.
+  ///
+  /// By default, this describes the results of the search in order to play
+  /// nicely with [expect] and its output when a failure occurs. If you wish
+  /// to get a string representation of the finder itself, pass [describeSelf]
+  /// as `true`.
+  @override
+  String toString({bool describeSelf = false}) {
+    if (describeSelf) {
+      return 'A finder that searches for ${TCandidate}s $requirements';
+    } else {
+      if (!hasFound) {
+        find();
+      }
+      return found.toString();
+    }
+  }
+}
+
+/// The results of searching with a [FinderBase].
+class FinderResult<TCandidate> extends Iterable<TCandidate> {
+  /// Creates a new [FinderResult] that describes the [values] using the given
+  /// [requirements].
+  ///
+  /// This description should be a brief English noun phrase describing the
+  /// finder's requirements. It should make sense as the end of the sentence
+  /// "Found an item with <description>".
```

Comment:
```suggestion
  /// "Found an item with <requirements>".
```

Although, as stated elsewhere I found the name "description" for this much better.